pymc/printing.py

```diff
@@ -130,148 +102,194 @@ def str_for_model(model: Model, formatting: str = "plain", include_params: bool
     def _expand(x):
         if x.owner and (not isinstance(x.owner.op, (RandomVariable, SymbolicRandomVariable))):
             return reversed(x.owner.inputs)
 
-    parents = [
+    return [
         x
         for x in walk(nodes=var.owner.inputs, expand=_expand)
         if x.owner and isinstance(x.owner.op, (RandomVariable, SymbolicRandomVariable))
     ]
-    names = [x.name for x in parents]
-
-    if "latex" in formatting:
-        return (
-            r"f("
-            + ",~".join([_latex_text_format(_latex_escape(n.strip("$"))) for n in names])
-            + ")"
-        )
-    else:
-        return r"f(" + ", ".join([n.strip("$") for n in names]) + ")"
 
 
-def _latex_text_format(text: str) -> str:
-    if r"\operatorname{" in text:
-        return text
+def _str_for_args_list(args: List[Variable], formatting: str) -> str:
+    """Create a human-readable string representation for the list of inputs
+    to a distribution or expression."""
+    strs = [_str_for_input_var(x, formatting=formatting) for x in args]
+    if formatting == "latex":
+        # Format the str as \text{} only if it hasn't been formatted yet and it isn't numeric
+        strs_formatted = [
+            s
+            if r"\text" in s or r"\operatorname" in s or s == "f()" or s.replace(".", "").isdigit()
+            else _latex_clean_command(s, command="text")
+            for s in strs
+        ]
+        return ",~".join(strs_formatted)
+    else:
+        return ", ".join(strs)
+
+
+def _latex_clean_command(text: str, command: str) -> str:
+    r"""Prepare text for LaTeX and maybe wrap it in a \command{}."""
+    text = text.replace("$", r"\$")
+    # str_for_model() uses \sim to format the array, and properly
+    # tilde in latex is hard. So we replace for simplicity
+    text = text.replace("~", "-")
+    if not text.startswith(rf"\{command}"):
+        # The printing module is designed such that text never passes through this
+        # function more than once. However, in some cases the text may have already
+        # been formatted for LaTeX -- such as when accessing a RandomVariable's
+        # pre-specified _print_name. In these cases we avoid the double wrap.
+        text = rf"\{command}{{{text}}}"
+    # This escape is a workaround for pymc#6508. MathJax is the latex engine in Jupyter
+    # notebooks, but its behavior in \text commands deviates from canonical LaTeX: stuff
+    # in the \text block will be rendered more or less verbatim. Meanwhile, other
+    # engines such as KaTeX behave differently and expect certain characters to be
+    # escaped to be typeset as desired. We work around this by escaping out of the
+    # command itself, writing the character, then continuing on with the same command.
+    if command == "text":
+        text = text.replace("_", rf"}}\_\{command}{{")
+    return text
+
+
+def _is_potential_or_deterministic(var: Variable) -> bool:
+    # This is a bit hacky but seems like the best we got
```

Comment:
Interesting function 😅. I never thought of checking if a variable is created via Potential or Deterministic this way. If it works, this sounds great but, as you suggest, it does look hackish...

```python
def _is_potential_or_deterministic(var: Variable) -> bool:
        try:
            return var.str_repr.__func__.func is str_for_potential_or_deterministic
        except AttributeError:
            # in case other code overrides str_repr, fallback
            return False
```

Reply:
Yep, I know I know....... :sweat_smile: 
I figured that it's not that far removed from the original implementation, since in the big-picture they're both inspecting whether a var's `str_repr` is a functools partial that was made in a certain way.

Anyways, a dispatch-based approach or at least requiring a model object would obviate this type of approach